### PR TITLE
feat: verify provenance before automatic updates

### PR DESCRIPTION
**Pull request by HastD**
Add an `ExecStartPre` item to `rpm-ostreed-automatic.service` that verifies build provenance for the image to be pulled.

Technically this is subject to a TOCTOU issue since a new image could be built after provenance verification is run and before the update itself starts. However, this is a very small window of time, and it would be difficult to avoid this without integrating provenance verification into `rpm-ostree` itself.

### CODE REVIEW

File: files/system/usr/libexec/secureblue/verify-provenance.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+set -euo pipefail
+
+image_ref=$(rpm-ostree status --booted --json | jq -cr '.deployments[0]."container-image-reference"')
+image_ref=${image_ref#*:docker://}
+echo "Verifying build provenance for ${image_ref}..."
+
+digest=$(crane digest "${image_ref}")
+echo "Image digest: ${digest}"
+
+slsa-verifier verify-image --source-uri 'github.com/secureblue/secureblue' "${image_ref}@${digest}"
```
Comment (RoyalOughtness) [inline files/system/usr/libexec/secureblue/verify-provenance.sh]: this should have source-branch set to `live`, otherwise it will validate against any branch

**Comment by RoyalOughtness** [comment]
we should probably test this in staging a bit due to Murphy's law. If there are edge cases that cause this to fail, it will be painful for users to get around it (rollback then upgrade after we revert)

**Comment by HastD** [comment]
I revised it to handle `br-*` tags as well as `latest` and to select the appropriate source branch in both cases. It will now bail out with a printed warning and exit code 0 if the booted image reference is not as expected (i.e. not an official secureblue image or the tag isn't `latest` or `br-*`).

As for ensuring there aren't edge cases that cause this to fail, instead of pushing it to staging, what do you think of the idea of initially setting it to send a notification to the user *without* canceling the update if verification fails? That way we'd hopefully get reports of it if provenance verification is failing for some users.

**Comment by HastD** [comment]
Actually, I just realized, even if this does fail in some edge case, it wouldn't be *that* annoying for users to work around, because it only affects *automatic* updates—manual updates with `rpm-ostree upgrade` would bypass provenance verification. The risk would be if users don't realize updates are failing and end up with an outdated image, though even in that case, the impact is limited by the existing service that notifies the user if their image is over a week out of date.

File: files/system/usr/libexec/secureblue/verify-provenance.sh [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+image_ref=$(rpm-ostree status --booted --json | jq -cr '.deployments[0]."container-image-reference"')
+image_ref=${image_ref#*:docker://}
+case "${image_ref}" in
+    ghcr.io/secureblue/*)
+        source_uri='github.com/secureblue/secureblue'
+        ;;
+    *)
+        echo "WARNING: Unknown image reference '${image_ref}'; unable to check provenance."
+        exit 0
```
Comment (RoyalOughtness) [inline files/system/usr/libexec/secureblue/verify-provenance.sh]: shouldn't this be a failure state? if someone isn't on secureblue / is on a fork, they need to update or disable this.

File: files/system/usr/libexec/secureblue/verify-provenance.sh [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    latest)
+        branch='live'
+        ;;
+    br-*)
+        branch="${image_tag#br-}"
+        branch="${branch%-*}"
+        ;;
+    *)
+        echo "WARNING: Unknown image tag '${image_tag}'; unable to check provenance."
+        exit 0
```
Comment (RoyalOughtness) [inline files/system/usr/libexec/secureblue/verify-provenance.sh]: this should also be a failure state imo

File: files/system/usr/libexec/secureblue/verify-provenance.sh (hunk 2 above)
Comment (HastD) [reply files/system/usr/libexec/secureblue/verify-provenance.sh]: That's a good point, I suppose that would be an unsupported configuration and it's on the user to modify or disable provenance verification to make it work with their setup.

**Review by RoyalOughtness — DISMISSED** [review]

**Review by RoyalOughtness — APPROVED** [review]

**Review by EsseLowNitro — APPROVED** [review]
